composer/callbacks/optimizer_monitor.py

```diff
@@ -0,0 +1,115 @@
+    | ``layer_second_moment_l2_norm/LAYER_NAME``    | is ``True``. Default: ``False``.                    |
+    |                                               |                                                     |
+    +-----------------------------------------------+-----------------------------------------------------+
+    |                                               | Layer-wise L2 norms of the step                     |
+    | ``layer_step_norm/LAYER_NAME``                |                                                     |
+    |                                               |                                                     |
+    +-----------------------------------------------+-----------------------------------------------------+
+    |                                               | Layer-wise cosine between the gradient and the step |
+    | ``layer_step_grad_cosine/LAYER_NAME``         |                                                     |
+    |                                               |                                                     |
+    +-----------------------------------------------+-----------------------------------------------------+
+    |                                               | Layer-wise ratio between step size and parameter    |
+    | ``layer_step_param_norm_ratio/LAYER_NAME``    | norm                                                |
+    |                                               |                                                     |
+    +-----------------------------------------------+-----------------------------------------------------+
+
+    Args:
+        log_layer_grad_norms (bool, optional): Whether to log the L2 normalization of each layer.
+            Default: ``False``.
+        log_layer_grad_norms (bool, optional): Whether to log optimizer-specific metrics.
+            Default: ``False``.
+    """
+
+    def __init__(self, log_layer_grad_norms: bool = False, log_optimizer_metrics: bool = False):
+        self.log_layer_grad_norms = log_layer_grad_norms
+        self.log_optimizer_metrics = log_optimizer_metrics
+
+    def batch_end(self, state: State, logger: Logger):
+        norm = 0.0
+        layer_norms = {}
+        optimizer_metrics = {}
+
+        for name, p in state.model.named_parameters():
+            if p.grad is not None and p.requires_grad:
+                param_grad_norm = p.grad.detach().data.norm(2).item()  # type: ignore
+                if self.log_layer_grad_norms:
+                    layer_norms[f'layer_grad_l2_norm/{name}'] = param_grad_norm
+
+                param_grad_norm = param_grad_norm**2
+                norm += param_grad_norm
+                metric_reporter = getattr(state.optimizers[0], 'report_per_parameter_metrics', None)
+                if callable(metric_reporter) and self.log_optimizer_metrics:
+                    optimizer_metrics = metric_reporter(p, name, optimizer_metrics)
```

Comment:
Right here is where we could add a prefix `optimizer_monitor/` to all metric keys

Reply:
I tried this out but the UX ended up feeling a little off having the default metrics (grad norms) separated from the optimizer metrics (moment norms etc.) since they are all conceptually related. I did end up following your suggesting of grouping the metrics hierarchically by type!